feat(epochs): backport the module from SDK

MaxMustermann2 opened the pull request:
This PR is a backport of the Cosmos SDK's epochs module. The backport allows us the flexibility to define our own epochs as default epochs and fixes a bug wherein the first epoch in the hook was always 2.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Introduced CLI query commands for the `epochs` module.
  - Added functionality for initializing and exporting genesis state for epochs.
  - Introduced new test helpers for setting up epoch-related queries and clients.

- **Bug Fixes**
  - Fixed import path issues related to the `epochs` package across multiple files.

- **Documentation**
  - Updated `x/epochs/README.md` to reflect new functionalities and changes.

- **Refactor**
  - Modified various import paths and package names to align with the new repository structure.

- **Tests**
  - Added new test cases to validate epoch increments, chain downtime handling, and querying epoch information.

- **Chores**
  - Updated workflow configurations to include necessary tokens and conditional setups.
  - Added exclusion patterns in `.semgrepignore` for specific simulation directories.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

MaxMustermann2 committed:
feat(epochs): backport the epochs module

MaxMustermann2 committed:
Merge develop

MaxMustermann2 committed:
fix: run go mod tidy

MaxMustermann2 committed:
fix(local): minute-wise epochs

MaxMustermann2 committed:
fix(dogfood): set default unbonding duration 7 day

MaxMustermann2 committed:
fix(local): unbond at 2 epochs

MaxMustermann2 committed:
fix(ci): fix failing jobs

MaxMustermann2 commented on x/epochs/keeper/epoch_infos.go:
```diff
@@ -0,0 +1,91 @@
+	return epoch, true
+}
+
+// setEpochInfoUnchecked sets the provided epoch info in the store, indexed by the identifier.
+// It performs no validation; the caller must ensure that it is valid and all the fields
+// are populated correctly.
+func (k Keeper) setEpochInfoUnchecked(ctx sdk.Context, epoch types.EpochInfo) {
+	store := ctx.KVStore(k.storeKey)
+	key := types.KeyEpoch(epoch.Identifier)
+	bz := k.cdc.MustMarshal(&epoch)
+	store.Set(key, bz)
+}
```
Is the comment not enough?

```
// It performs no validation; the caller must ensure that it is valid and all the fields
// are populated correctly.
```

MaxMustermann2 commented on x/epochs/keeper/keeper.go:
```diff
@@ -0,0 +1,55 @@
+	}
+)
+
+// NewKeeper creates a new epochs keeper. it is returned as a pointer since
+// the primary purpose of an epochs keeper to a caller is to use its hooks.
+// the hooks should be set on the pointer for them to take effect.
+func NewKeeper(
+	cdc codec.BinaryCodec,
+	storeKey storetypes.StoreKey,
+) *Keeper {
+	return &Keeper{
+		cdc:      cdc,
+		storeKey: storeKey,
+	}
```
The hooks object is accessed through the `Hooks()` function, which has been modified to take care of this case. 

```go
// Hooks returns the hooks registered to the module.
func (k Keeper) Hooks() types.EpochHooks {
	if k.hooks != nil {
		return k.hooks
	}
	return types.MultiEpochHooks{}
}
```

MaxMustermann2 committed:
refactor(epochs): respond to ai review

MaxMustermann2 commented on x/epochs/client/cli/query.go:
```diff
@@ -0,0 +1,109 @@
+
+// GetCmdEpochsInfos returns a command for querying all the currently available epoch infos.
+func GetCmdEpochsInfos() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "epoch-infos",
+		Short: "Query all epochInfos",
+		Example: strings.TrimSpace(
+			fmt.Sprintf(`$ %s query %s epoch-infos`,
+				version.AppName, types.ModuleName,
+			),
+		),
+		Args: cobra.NoArgs,
+		RunE: func(cmd *cobra.Command, _ []string) error {
+			clientCtx, err := client.GetClientQueryContext(cmd)
+			if err != nil {
+				return err
+			}
+			queryClient := types.NewQueryClient(clientCtx)
+
+			pageReq, err := client.ReadPageRequest(cmd.Flags())
+			if err != nil {
+				return err
+			}
+
+			req := &types.QueryEpochsInfoRequest{
+				Pagination: pageReq,
+			}
+
+			res, err := queryClient.EpochInfos(cmd.Context(), req)
+			if err != nil {
+				return err
+			}
+
+			return clientCtx.PrintProto(res)
+		},
+	}
+
+	flags.AddQueryFlagsToCmd(cmd)
+
+	return cmd
```
`PrintProtoLegacy`  was replaced with `PrintProto. LGTM.

MaxMustermann2 commented on x/epochs/README.md:
```diff
@@ -0,0 +1,84 @@
+This contains the current state of the timer with the corresponding identifier.
+Its fields are modified at every timer tick.
+EpochInfos are initialized as part of genesis initialization or upgrade logic,
+and are only modified during the begin blocker.
```
`begin blocker` is a Cosmos term.

MaxMustermann2 commented on proto/exocore/epochs/v1/genesis.proto:
```diff
@@ -0,0 +1,67 @@
+  string identifier = 1;
+  // start_time of the epoch. it may be in the future, in which case, the epoch
+  // is not yet active.
+  google.protobuf.Timestamp start_time = 2
```
It's in the next line because the line hasn't ended and there is more information after the `2`

MaxMustermann2 committed:
fix(docs): add `the` before module

Co-authored-by: coderabbitai[bot] <136622811+coderabbitai[bot]@users.noreply.github.com>

MaxMustermann2 committed:
Merge develop

MaxMustermann2 commented on x/epochs/keeper/abci.go:
```diff
@@ -0,0 +1,104 @@
+					),
+				)
+				// NOTE: this hook is called BEFORE the new epoch info is saved.
+				k.Hooks().AfterEpochEnd(ctx, epochInfo.Identifier, epochInfo.CurrentEpoch)
```
In the previously imported implementation, the `AfterEpochEnd` function was called after incrementing the `epochInfo.CurrentEpoch`, resulting in the same epoch (bearing the increased number) ending and then starting immediately afterwards

MaxMustermann2 replied:
As a result of this bug, the effective opt-out completion epoch was reduced by 1. 

For example, with the number of unbonding epochs set to 2, if an operator opts out during epoch 5, they will stop validating at the end of epoch 5 and complete the opt out at the end of epoch 7. This is accurately scheduled with or without the bug being present in the `x/epochs` module, because the currently running epoch number is accurately fetched during the time of opt out.

With the bug, however, the epoch reported in the `AfterEpochEnd` hook is 7 even though the epoch currently ending is 6. The hook then finds all the opt-outs completing at epoch 7 (instead of epoch 6) and schedules them to be removed at the end of the block.

MaxMustermann2 replied:
Of course, we can work around this bug by first decrementing the `currentEpoch` variable within each hook. But I felt that it's better to bring ownership of this code under our primary repository, given that a lot of our design depends on the epochs system.

cloud8little commented:
Query the epoch information with two successive command `exocored q epochs epoch-infos`,  look into the `minute ` epoch information, the current_epoch increase by 1 while the time latency for two successive command is obivioulyless than one minute. and the `current_epoch_start_time` in the minute identifier is incorrect. 

NOTE: the genesis_time is "2024-06-14T08:34:18.95676475Z"

```
[root@testnet-exochain testscript]# exocored q epochs epoch-infos
epochs:
- current_epoch: "3"
  current_epoch_start_height: "3"
  current_epoch_start_time: "2024-06-16T08:34:18.956764750Z"
  duration: 86400s
  epoch_counting_started: true
  identifier: day
  start_time: "2024-06-14T08:34:18.956764750Z"
- current_epoch: "68"
  current_epoch_start_height: "68"
  current_epoch_start_time: "2024-06-17T03:34:18.956764750Z"
  duration: 3600s
  epoch_counting_started: true
  identifier: hour
  start_time: "2024-06-14T08:34:18.956764750Z"
- current_epoch: "423"
  current_epoch_start_height: "423"
  current_epoch_start_time: "2024-06-14T15:36:18.956764750Z"
  duration: 60s
  epoch_counting_started: true
  identifier: minute
  start_time: "2024-06-14T08:34:18.956764750Z"
- current_epoch: "1"
  current_epoch_start_height: "1"
  current_epoch_start_time: "2024-06-14T08:34:18.956764750Z"
  duration: 604800s
  epoch_counting_started: true
  identifier: week
  start_time: "2024-06-14T08:34:18.956764750Z"
pagination:
  next_key: null
  total: "4"
[root@testnet-exochain testscript]# exocored q epochs epoch-infos
epochs:
- current_epoch: "3"
  current_epoch_start_height: "3"
  current_epoch_start_time: "2024-06-16T08:34:18.956764750Z"
  duration: 86400s
  epoch_counting_started: true
  identifier: day
  start_time: "2024-06-14T08:34:18.956764750Z"
- current_epoch: "68"
  current_epoch_start_height: "68"
  current_epoch_start_time: "2024-06-17T03:34:18.956764750Z"
  duration: 3600s
  epoch_counting_started: true
  identifier: hour
  start_time: "2024-06-14T08:34:18.956764750Z"
- current_epoch: "424"
  current_epoch_start_height: "424"
  current_epoch_start_time: "2024-06-14T15:37:18.956764750Z"
  duration: 60s
  epoch_counting_started: true
  identifier: minute
  start_time: "2024-06-14T08:34:18.956764750Z"
- current_epoch: "1"
  current_epoch_start_height: "1"
  current_epoch_start_time: "2024-06-14T08:34:18.956764750Z"
  duration: 604800s
  epoch_counting_started: true
  identifier: week
  start_time: "2024-06-14T08:34:18.956764750Z"
pagination:
  next_key: null
  total: "4"
```

```
[root@testnet-exochain testscript]# date
Mon Jun 17 04:09:16 UTC 2024
[root@testnet-exochain testscript]# date
Mon Jun 17 04:09:18 UTC 2024
[root@testnet-exochain testscript]# date
Mon Jun 17 04:09:20 UTC 2024
```

MaxMustermann2 committed:
epochs: return block time in query

The returned block time can be used to determine whether any of the
epochs are about to change.

MaxMustermann2 committed:
doc(epochs): clarify no problems with timezones

MaxMustermann2 commented:
> NOTE: the genesis_time is "2024-06-14T08:34:18.95676475Z"

Given that is the case, this behavior is expected. It is the same functionality as the one implemented in the Cosmos SDK and the previously imported module.

The current epoch, given any identifier, indicates the number of periods of that epoch since the epoch counting start time (which defaults to the genesis time, if not set). For example, if the day epoch is reported as 3, it means that the 3rd day (since counting started) is ongoing.

However, the number is incremented only once during a block. This is an intentional design, since doing otherwise would result in epochs being skipped. This is the cause for the deviation that you observed.

Starting a chain with a genesis time in the past causes each produced block to increase the current epoch number by 1 until the epoch's scheduled end time (= current start time + duration) < block time.

In the provided example, the chain's genesis time was 2 days, 19 hours, 32 minutes and ~16 seconds ago. The start_time for each epoch is set to the genesis time.

Since the start_time, the 3rd day has started, which is accurately reflected in your query.
Similarly, it is the 68th hour that is currently ongoing, which is also correctly reported.
The week is the first week since the genesis time, which is again accurate.

The minute, however, is increasing by 1 at each block. 4052.3 minutes have completed between the genesis time and now, so it will increase till 4053 blocks (+ some extra for the time it takes for block production to get there) and then continue on its regular minute-wise schedule.

The comments in the code describe this detail as well.
https://github.com/ExocoreNetwork/exocore/blob/91a9681fd5448474b76229e6ef3b7193107fcc32/proto/exocore/epochs/v1/genesis.proto#L33-L51

MaxMustermann2 committed:
fix(ci): provide github token to buf setup

avoid rate limits

MaxMustermann2 committed:
feat(ci): allow super-linter running with act

Example usage:
act --workflows .github/workflows/super-linter.yml pull_request

MaxMustermann2 committed:
Merge with `develop`

MaxMustermann2 committed:
remove unused file

MaxMustermann2 committed:
fix: typo in script

MaxMustermann2 commented:
@coderabbitai review

bwhour commented on x/epochs/keeper/abci.go:
```diff
@@ -0,0 +1,113 @@
+				// short circuit if this epoch is not yet scheduled to start
+				return false
+			}
+			epochEndTime := epochInfo.CurrentEpochStartTime.Add(epochInfo.Duration)
```
can  this  epochInfo.Duration be zero ?  Any problems if it is zero ?

MaxMustermann2 replied:
It can be zero. It will result in the epoch number increasing by one at each block. I'll test and report back.

MaxMustermann2 replied:
Sorry, my bad. The epoch duration cannot be zero. It is validated in the genesis (stateless) validation.
https://github.com/ExocoreNetwork/exocore/blob/53863434195b96fa51a2fb6f0ad7f000ff86b24a/x/epochs/types/genesis.go#L48-L50

bwhour commented on x/epochs/keeper/epoch_infos.go:
```diff
@@ -0,0 +1,92 @@
+	ctx sdk.Context, identifier string,
+) (epoch types.EpochInfo, found bool) {
+	store := ctx.KVStore(k.storeKey)
+	key := types.KeyEpoch(identifier)
```
this two lines can be abstract into one function as it appears too many times.

> `func (k Keeper) getStoreAndKey(ctx sdk.Context, identifier string) (store sdk.KVStore, key []byte)`

MaxMustermann2 replied:
I'll change it to the prefix store instead to accomodate your opinion.

MaxMustermann2 replied:
Actioned in 3cda16e8

bwhour commented on x/epochs/types/identifier.go:
```diff
@@ -0,0 +1,43 @@
+// ValidateEpochIdentifierString accepts a string and validates it as an epoch identifier.
+// It is not used directly by this module; rather it is created for other modules to validate
+// their params.
+func ValidateEpochIdentifierString(s string) error {
```
The current ValidateEpochIdentifierString function only checks that the string is not null, it does not verify that the string is one of the defined constants. An additional check can be added to ensure that the string is a valid epoch identifier

MaxMustermann2 replied:
The epoch identifier can be any arbitrary string. The constants that we have defined only refer to the default epochs. For example, it is possible to set up an identifier (within the genesis file) called "half-hour" with duration of 30 minutes (or even 20 minutes for that matter; the name and the duration aren't validated).

MaxMustermann2 committed:
refactor(epochs): use prefix store to DRY

MaxMustermann2 commented:
Test object: 3cda16e8

Results for minute epoch
```
$ exocored query epochs current-epoch minute --output json | jq -r .current_epoch && sleep 60 && exocored query epochs current-epoch minute --output json | jq -r .current_epoch
2
3
```

Associated logs
```
2:07PM INF ending epoch identifier=minute module=x/epochs number=2
```

MaxMustermann2 committed:
Merge upstream/develop

MaxMustermann2 committed:
fix(avs): use local epochs module

MaxMustermann2 committed:
fix(dep): update vuln dep

MaxMustermann2 committed:
fix(epochs): add extra validations

bwhour reviewed (approved):
it looks good.
